### DISCUSSION TITLE
Framerate dropped from 10fps to 1fps when changing the position

### DISCUSSION THREAD

**Discussion by kohillyang** [pixijs/pixijs#7937] [General]
Hello, I'm using pixi.js to show a huge image(20k x 120k), the image is huge so I can't store it by a single Texture. Instead, I calculate the pixels in each frame using a function called wrapAffine. That function is written in CPP with CEF(chrome embedding framework) and it took only about 1ms to wrap a 20k image to 1k. When the user scrolls its mouse wheel or drags the image. The image should scale or move. The example code is in <https://codepen.io/kohillyang/pen/xxLXrwg>. Please note that warpaffinePixi provided in this code is much slower than the CPP version. But it does not affect the final conclusion.

If you are using Edge `95.0.1020.40`, and just run the code in <https://codepen.io/kohillyang/pen/xxLXrwg>.
As you can see in <https://codepen.io/kohillyang/pen/xxLXrwg>. When the user drags the image using the right mouse, the image is moving. And the framerate is about 10fps on my laptop. however, if your Devtools is Open, then it lags. I can not see the framerate because the framerate shown by Edge is wrong.

If you are using Edge `95.0.1020.40`, and just copy the code, create an index.html, and run it locally.
![GIF 2021-10-31 21-22-17](https://user-images.githubusercontent.com/7751507/139585664-005cf379-810c-4e8a-9993-a5e5970c9d2c.gif)
the framerate is normal when trying to scale the image, but it's too low when dragging. And if you are recording in Devtools' performance window, then it becomes normal.

If you are using Firefox 92.0.1
The code cannot handle the mouse wheel event. And the framerate is normal when I try to drag the image

- **`pixi.js` version**:  The lastest version.
- **Browser & Version**: Edge `95.0.1020.40`, 
- **OS & Version**: _e.g. windows 10
- **Running Example**: <https://codepen.io/kohillyang/pen/xxLXrwg>

**Comment by kohillyang** [comment]
I can not understand why the framerate of dragging is lower than scaling. Does it because of the Cache strategy? Or why in fireFox it is OK. Does it mean that WebGL performance in firefox is better than in Chrome?  This seems to be a little strange to me.

**Comment by eXponenta** [comment]
Because you every changes enumerate all million pixels in for loop and then send texture to GPU with resizing. CPU -> GPU sync is very expensive and for 4k images can be about 100 ms on low tire devices.

This is not pixi problem.

@bigtimebuddy move this to discussion .

**Comment by kohillyang** [comment]
@eXponenta A 1000x1000k image is only 1MB. Considering a GPU whose bandwidth is 1GB/s, which is slow enough, and uploading that image to the GPU will only take less than 1ms. I know you would say that the draw call is slow. But will a draw call take me over 100ms?  I can't believe that.

**Reply by eXponenta** [reply]
Bandwidth and upload time not same.
Before push data to texture, this texture should be locked (mapped) and CPU->GPU task runned, taht requires a some synchronization CPU and GPU, this is not dependent of bandwidth, this depending of how many times GPU is stall, because there are a lot of another processes, that can use GPU too, include browser, this time is complitenly more that 1/1GB/s.
Than you should know that a WebGL is abstractions on ANGLE/Metal that has a huge overhead.

Than, @photonstorm noticed - you call regeration every mouse events, but should only mark latest data and call update before render if position/scale was changed.

**Comment by photonstorm** [comment]
It's not 1MB of GPU memory and unless you're uploading compressed / packed texture data, you'll pay the conversion cost as well. Assuming it does take 1ms to upload, do that 60 times in a second, and your frame budget is already blown.

**Comment by kohillyang** [comment]
@photonstorm I don't understand what you mean. Assuming it does take 1ms to upload, then I should get 1000fps, doesn't it? But the real fps is less than 10fps. and Since Dragging and scaling should spend equal time to upload the image to GPU, then their fps should be the same. What seems to be strange is that the fps of dragging is much lower than scaling.

**Comment by kohillyang** [comment]
And I also do not understand what you said `It's not 1MB of GPU memory and unless you're uploading compressed`. A 1k x1k unsigned byte gray image will take me over 1MB of GPU memory? OK, let assume it is a rgba 24bit image. Then Sending it to the GPU would at most take me 4ms. Then I should get 250fps, not less than 10fps.

**Comment by photonstorm** [comment]
Mouse wheel polling rates (depending on hardware and OS) can fire hundreds of times per second. Even a 500Hz mouse will trigger 500 updates. My mouse is 1000Hz. Not all of these polls will manifest as browser events, of course, but it can still be a massive number if you're not careful (or rate-limit it in your code). Add a 4ms texture cost on top of that and your frame rate is gone.

**Comment by kohillyang** [comment]
I agree with what @photonstorm said. And after looking at the sources of Pixiv.js, I found I can write the new image data into the original buffer without creating a new Texture. Also, to avoid triggering too many updates, I use a dirty flag to indicate whether the sprite needs to be updated. Now the bottleneck is the function warpaffinePixi. There is a huge gap between v8 and CPP. CPP version only needs less than 1ms on my laptop, but v8 need over 40ms. 


Sources are updated in <https://codepen.io/kohillyang/pen/xxLXrwg>, I hope it is useful to somebody.

**Reply by photonstorm** [reply]
Sounds like it could be an ideal candidate for wasm / web workers.

**Comment by kohillyang** [comment]
@photonstorm I found the canvas becomes white/black when the size of the canvas changes, do you know how to solve this? Thanks.

**Reply by photonstorm** [reply]
Depends on how you're resizing the canvas. Pixi has features built-in to let you resize the renderer, which you'd need to do if you're resizing the canvas outside of it.

**Reply by kohillyang** [reply]
I mean the whole canvas is filled with black. The document says `Hint: All DisplayObjects (i.e. Sprites) that render to a RenderTexture should be preloaded otherwise black rectangles will be drawn instead.` in <https://pixijs.download/dev/docs/PIXI.RenderTexture.html>. So this may be because I created a texture but did not load it, then a black rectangle was drawn. But I do not know how to load it.

**Reply by kohillyang** [reply]
I found the reason is that I must call app.render() since app.renderer.resize will clear the canvas.
